pithos/plugins/notification_icon.py

```diff
@@ -12,57 +12,142 @@
     description = 'Adds pithos icon to system tray'
 
+    appindicator = None
+
     def on_prepare(self):
-        if indicator_capable:
-            self.ind = AppIndicator.Indicator.new("io.github.Pithos-tray",
-                                                  "io.github.Pithos-tray",
-                                                  AppIndicator.IndicatorCategory.APPLICATION_STATUS)
-            # FIXME: AppIndicator might be falling back to XEmbed
-        elif not backend_is_supported(self.window):
-            self.prepare_complete(error='Notification icon requires X11 or AppIndicator')
-        else:
+        # 1st we check for AppIndicator3.
+        # If it's not found we check to make sure we aren't running on KDE Plasma
+        # because it doesn't support Gtk.StatusIcon. If we aren't we then
+        # check to make sure we aren't running in Wayland because GNOME Shell doesn't support
+        # tray icons in Wayland.
+        # If AppIndicator3 is found we check for com.canonical.indicator.application
+        # or if we have the appindicatorsupport GNOME Shell extension installed and enabled,
+        # if not we check to see if we're running in KDE Plasma. If we aren't running in KDE Plasma,
+        # have the indicator.application service or have the appindicatorsupport GNOME Shell extension
+        # installed and enabled we have to check to see if we're on Wayland to see if we can use Gtk.StatusIcon.
+        # It's all rather convoluted...
+        def check_for_x():
+            self.appindicator = None
+            display = self.window.props.screen.get_display()
+            if not type(display).__name__.endswith('X11Display'):
+                logging.debug('Notification icon requires X11 or AppIndicator Support')
+                self.prepare_complete(error='Notification icon requires X11 or AppIndicator Support')
+            else:
+                self.prepare_complete()
+
+        def init_appindicator():
+            self.ind = self.appindicator.Indicator.new(
+                "pithos-tray-icon",
+                "pithos-tray-icon",
+                self.appindicator.IndicatorCategory.APPLICATION_STATUS,
+            )
+
             self.prepare_complete()
 
-    def on_enable(self):
-        self.delete_callback_handle = self.window.connect("delete-event", self._toggle_visible)
-        self.state_callback_handle = self.window.connect("play-state-changed", self.play_state_changed)
-        self.song_callback_handle = self.window.connect("song-changed", self.song_changed)
+        def has_appindicator_shell_ext():
+            schema = Gio.SettingsSchemaSource.get_default()
+            valid_schema = schema.lookup('org.gnome.shell', True)
+            if valid_schema:
+                gnome_shell = Gio.Settings.new('org.gnome.shell')
+                enabled_extensions = gnome_shell.get_value('enabled-extensions')
+                if 'appindicatorsupport@rgcjonas.gmail.com' in enabled_extensions:
```

Comment:
I don't think this is the right way, we should be checking for the kde bus name it owns.